woocommerce/woocommerce-gutenberg-products-block
# Enable more style options for the Featured Category block

This PR adds several style options to the Featured Product block. The current status of features supported and planned for each block can be seen in [this spreadsheet](https://docs.google.com/spreadsheets/d/1KsKoq3J86s4PRZoHufj_qshGl_tTWjfolJXsrqE2AsU/edit#gid=1253950386).

**This PR fixes the following problems:**

* Previously, the overlay of this block was actually rendered as a background. That meant inconsistent behavior in certain circumstances (like when an external padding was added). The overlay is now a separate DOM element.
* Previously, the overlay color was controlled within the `Color` section of the sidebar, while its opacity was controlled elsewhere. This PR consolidates the controls under the `Overlay` section.
* Previously, the featured product image was rendered as a background. This meant we could not correctly apply things like duotone filters (see #5555). The image is now rendered as a separate DOM element.
* Previously, the featured image was resized constraining it to the block size. This meant that the focal point could only be moved vertically. Now the focal point can be moved correctly (see below: **Image fit options**).
* The resizable container handle was not completely visible (it would only show half a circle). This is fixed.
* The resizable container handle was always visible regardless of the selected status of the block.
* Previously, if the padding was large enough, the content could be pushed out of the container. Now the block behaves more like the Cover block, in the sense that the padding will scale the container up. This also means that the resizable control is subject to this rule, and can't resize the container lower than a certain minimum (determined by content and size). This is the same behavior as the Cover block.

**This PR adds the following features:**

* **Support for Duotone:** users can now set a duotone filter on the featured product image.
* **Support for gradients:** users can now set the overlay color as a gradient.
* **Support for padding:** users can now set custom padding for the block.
* **Image fit options:** The user can now decide how should the image behave on the resizing of the component; it can either scale to always cover the entire container, or remain its original size.


Closes #6151

### Screenshots

![Screen Shot 2022-04-16 at 21 21 49](https://user-images.githubusercontent.com/1847066/163688658-416fb46c-7c4b-478d-904b-89d9b0d649a5.png)

### Testing

#### Automated Tests

### Manual Testing

How to test the changes in this Pull Request:

1. Use a block theme (such as Twenty Twenty Two)
2. Add the ”Featured Category” block to a page and select a product.
3. Make sure that the resizable handle is a shown fully (it should be a full circle), and that is shown only when the block is selected.
4. Select the block and apply a Duotone filter.
5. Make sure the overlay color control is under the `Overlay` section and change the overlay color to a gradient.
6. Add a padding to the block and make sure the padding is added on the inside of the wrapper (i.e. no white space is created outside of the block).
7. Move around your image focal point and make sure you can actually move around the entire image and not only constrained on one dimension.
8. Apply “Image fit > Cover” and notice how the image will scale along with the container. As opposed to point 5 above, now the focal point picker is limited to one dimension.
9. Increase the `padding-top` of the block and notice that the content can't get pushed out of the container, but instead the container resizes.
10. When that's the case, try resizing the block through the handle to a lower height: the handle should move but the container should not resize. When the mouse is lifted, the handle should return to its original position.
11. For the steps 4–9 above, save the page and check that the styles are applied correctly on the front-end (as the block is rendered statically via PHP).

### User Facing Testing
These are steps for user testing (where "user" is someone interacting with this change that is not editing any code).
### Known issues

Currently, the `border` controls don't seem to work. This appears to be an issue related to the current Gutenberg version shipped with Core ([Slack thread](p1648672145909859-slack-C02SGH7JBGS)).

---

**❗  Note:** Since this PR includes several changes, [we agreed that it should have a multi-line changelog](p1649199937104809-slack-C02SGH7JBGS). I checked the automation script which creates the changelog, and it seems to be fine with that. There might be some formatting issues with newlines and such, so perhaps, _release lead_, you might have to adjust a few things before publishing.

### Changelog

>  Enhanced the *Featured Category block*:
> 
> * Implemented support for duotone.
> * Implemented support for gradients on the overlay.
> * Implemented support for custom inner padding.
> * Implemented image fit options: users can now decide how should the image behave on the resizing of the component; it can either scale to always cover the entire container, or remain its original size.
> * Fixed an inconsistency where the overlay color was controlled by the background color control. It is now moved to the correct section.
> * Fixed the focal point picker, it now works on both axis as long as the image fit (above) is set to `none`.
> * Fixed an issue with the visibility of the resizing handle.
> * Fixed an issue which would keep the resizing handle always active regardless of block selection status.
> * Changed the behavior of the resizing: the block can't be resized below a minimum height determined by its content plus the padding.
